### ISSUE TITLE
[Feature Request]: Add yt-dlp intergration

### ISSUE THREAD

**Issue by KevinReems** [FreeTubeApp/FreeTube#4178]
### Guidelines

- [X] I have searched the issue tracker for [open](https://github.com/FreeTubeApp/FreeTube/issues?q=is%3Aopen+is%3Aissue) and [closed](https://github.com/FreeTubeApp/FreeTube/issues?q=is%3Aissue+is%3Aclosed) issues that are similar to the feature request I want to file, without success.
- [X] I have searched the [documentation](https://docs.freetubeapp.io/) for information that matches the description of the feature request I want to file, without success.
- [X] This issue contains only one feature request.

### Problem Description

Currently you can only download a limited selection of video qualities with audio included.

### Proposed Solution

yt-dlp is capable of downloading and then muxing high resolution video with high bitrate audio streams.

For example: yt-dlp.exe -f bestvideo[height<=1080]+bestaudio will output a 1080p files with the best available audio stream available to download. Since nothing is actually encoded, this process is extremely fast.

A few menu options could be added to Settings to specify the path of yt-dlp, output directory, desired/maximum resolution etc.

### Alternatives Considered

There may be other command line downloaders but since yt-dlp seems to be the clear winner it would be best to start there.

### Issue Labels

support for external software

**Comment by absidue** [comment]
To avoid us having to implement loads of individual settings, especially considering the amount of possible options even just for yt-dlp's `-f` flag, I would suggest that this be implemented similar to the external player settings. One field to specify the path to yt-dlp, another field to set the command line options for yt-dlp, a link to yt-dlp's README so people can find out what options they can specify and we can reuse the existing download folder setting.

FreeTube would then call yt-dlp like this: `[path to binary] [your specified commandline options] https://www.youtube.com?v=[video id]`.

**Comment by Aphexus** [comment]
I'd only say it should be more general. Instead of working it as if one must yt-dlp, just call it "external downloader" or "custom downloader". One should be able to wrap any calls to yt-dlp in a batch file as long as they know the link. If batch downloading is added(such as downloading a watch list) then either this could be iterated or an additional parameter can be passed that signals it is a list(since yt-dlp allows for lists).

E.g., if one has a text field then one can just type in the command to execute: "C:\yt-dlp.exe -f best %1" or whatever and all freetube has to do is scan for the tokens to replace(e..g, %1 or maybe <url>).


The point is there is no reason to fix the downloader since in the future it might need change. Even if it is only in terms of wording. basically we need to just have an external downloader.

**Comment by absidue** [comment]
@Aphexus It's actually worse adding support for a generic one rather than a specific one. It would be much harder to integrate, because instead of being able to show useful error messages to users we would have to show the raw console output. As you brought up the topic of integration sich as downloading lists, things like that are significantly harder when you have to deal with unknown downloader. Additionally it would put a lot more support burden on the developers, as most users would struggle to set it up correctly, considering how many people already struggle with the external player settings.

A generic one is too much of a power user feature to add to FreeTube, I would suggest calling your batch scripts and different downloaders yourself.

**Comment by Aphexus** [comment]
> @Aphexus It's actually worse adding support for a generic one rather than a specific one. It would be much harder to integrate, because instead of being able to show useful error messages to users we would have to show the raw console output. As you brought up the topic of integration such as downloading lists, things like that are significantly harder when you have to deal with unknown downloader, it would also be impossible to show the download progress or cancel the downloads if a download manager were ever added. Additionally it would put a lot more support burden on the developers, as most users would struggle to set it up correctly, considering how many people already struggle with the external player settings.
> 
> A generic one is too much of a power user feature to add to FreeTube, I would suggest calling your batch scripts and different downloaders yourself.

This is absolutely wrong. If the entire point of your logic is "So we can show nice error messages in a nicely formatted box" then maybe. But you are removing tremendous power(which you agree) simply to give nice error messages. WHY? Why would you do that? Why would you sacrifice giving more power to the user so you can make a nice looking error output?

You don't have to deal with an unknown downloader. What your claim is "Do it my way" because you want it your way and you don't want to allow for other peoples methods(which happen to include your own).

You are simply wrong.

There are ways to handle your case AND what I ask for. To have the power AND specificity you want(but you want to limit the power I ask for).


1. Enabling running a script or app to download rather than forcing a specific downloader does not a priori limit one using a specific downloader(such as yt-dlp). There are two ways one can handle this: 1. Either separate them inside freetube(allow both) or  Detect if the app being used is yt-dlp then run all the custom code your method would have to require in the first place(which is actually more than mine.  Your method requires far more work than mine does.
2. running stuff in a batch file can still output nicely formated information. E.g., one could put it in a vbs script or use nircmd to put the error in a message box. One can make a custom downloader that wraps any other.

Again, your logic is "Do it might way because it's easier" when it is provably wrong and actually worse because it limits everyone to your method which actually may not work for everyone. What happen in 5 years say that yt-dlp is no longer maintained  and no longer works with youtube? Then someone has to make a request to remove the code and someone else has to make a request to add a new way to download(and if they are like you then they will create the same set of problems).

Please learn how generalization works and it's benefits. It will cause less problems in the world not more. If you get away from "My way is right and yours is wrong" it actually creates less problems not more like you believe. (note that my solution does not prevent your solution but your solution prevents mine(explicitly) hence why mine is better. Again, this is the power of generalization(generalization is not exclusive but inclusive and it does not prevent specificity like you believe it does).

**Comment by absidue** [comment]
@Aphexus You are welcome to create a custom build of FreeTube for yourself with as much customisation and power user tools as you want. FreeTube however is meant to be user friendly, not maximise the possibilities for the few while making it much harder to use for everyone else.

One of the most power user features in FreeTube is the external player functionality and a large amount of users already struggle with that, what you are suggesting would be significantly more powerful, so it follows that a lot more people would struggle with it and we would have to waste a lot more time providing support for them, which could easily be avoided by providing proper integration with a tool.

This has nothing to do with generalisation, it comes down to you wanting us to make your life better to the detriment of our time and other users.

**Comment by Aphexus** [comment]
@absidue  You are insane. You are illogical and your reasoning is a dead end. You claim I want to make my life better to the detriment of others but that is precisely you. Way to project.

1. My solution is easier to implement. So it actually costs less time than yours.
2. My solution is more general and applies to more people than just me. Yours only applies to those that use yt-dlp.

You are insane. I can see we will never come to any meaningful understanding.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
@Aphexus i advise you to stop being disrespectful towards Absidue
